crates/relay/src/lib.rs

```diff
@@ -241,13 +245,92 @@ pub mod pallet {
                 reports.insert(vault_id.clone());
             });
 
-            // reward the participant by increasing their SLA
+            // reward the reporter by increasing their SLA
             ext::sla::event_update_vault_sla::<T>(&signer, ext::sla::Action::TheftReport)?;
-
             Self::deposit_event(<Event<T>>::VaultTheft(vault_id, tx_id));
 
             Ok(().into())
         }
+
+        /// Report Vault double payment, providing two fraud proofs (malicious Bitcoin transactions
+        /// and the corresponding transaction inclusion proofs).
+        ///
+        /// # Arguments
+        ///
+        /// * `origin`: Any signed user.
+        /// * `vault_id`: The account of the vault to check.
+        /// * `raw_merkle_proofs`: The proofs of tx inclusion.
+        /// * `raw_txs`: The raw Bitcoin transactions.
+        #[pallet::weight(<T as Config>::WeightInfo::report_vault_theft())]
+        #[transactional]
+        pub fn report_vault_double_payment(
+            origin: OriginFor<T>,
+            vault_id: T::AccountId,
+            raw_merkle_proofs: (Vec<u8>, Vec<u8>),
+            raw_txs: (Vec<u8>, Vec<u8>),
+        ) -> DispatchResultWithPostInfo {
+            ext::security::ensure_parachain_status_not_shutdown::<T>()?;
+            let signer = ensure_signed(origin)?;
+
+            ensure!(
+                raw_merkle_proofs.0 != raw_merkle_proofs.1,
+                Error::<T>::DuplicateMerkleProof,
+            );
+            ensure!(raw_txs.0 != raw_txs.1, Error::<T>::DuplicateTransaction);
+
+            let (left_proof, right_proof) = match raw_merkle_proofs {
+                (left, right) => (
+                    ext::btc_relay::parse_merkle_proof::<T>(&left)?,
+                    ext::btc_relay::parse_merkle_proof::<T>(&right)?,
+                ),
+            };
+
+            let (left_tx, right_tx) = match raw_txs {
+                (left, right) => (
+                    parse_transaction(left.as_slice()).map_err(|_| Error::<T>::InvalidTransaction)?,
+                    parse_transaction(right.as_slice()).map_err(|_| Error::<T>::InvalidTransaction)?,
+                ),
+            };
+
+            let left_tx_id = left_tx.tx_id();
+            let right_tx_id = right_tx.tx_id();
+
+            // ensure both transactions are included
+            ext::btc_relay::verify_transaction_inclusion::<T>(left_tx_id, left_proof)?;
+            ext::btc_relay::verify_transaction_inclusion::<T>(right_tx_id, right_proof)?;
+
+            let vault = ext::vault_registry::get_active_vault_from_id::<T>(&vault_id)?;
+            ensure!(
+                Self::has_input_from_wallet(&left_tx, &vault.wallet)
+                    && Self::has_input_from_wallet(&right_tx, &vault.wallet),
+                Error::<T>::VaultNoInputToTransaction
+            );
+
+            match (
+                OpReturnPaymentData::<T>::try_from(left_tx),
```

Comment:
Could use question mark operator instead of match to save some horizontal space

Reply:
I quite like this construction, not sure it would look better with the question mark operator since we would need to assign both tuple fields to a local variable.